File: app/backend/src/llm/proxyAdmin.ts
// Typed client for the CLIProxyAPI Management API (/v0/management).
// Lets Dana add/edit/remove custom upstream providers (OpenAI-compatible and
// Anthropic-compatible) by API key + base URL — e.g. MiniMax — at runtime.
// CLIProxyAPI persists changes to its config.yaml and hot-reloads (no restart).
//
// Verified contract (against eceasy/cli-proxy-api:latest):
//   Base:  $PROXY_BASE_URL/v0/management
//   Auth:  Authorization: Bearer <secret-key>   (missing → 401 "missing management key")
//   GET    /openai-compatibility → { "openai-compatibility": [...] | null }
//   PUT    /openai-compatibility → body is the raw array (replaces whole list) → { status: "ok" }
//   DELETE /openai-compatibility?name=x → { status: "ok" }
//   GET/PUT /claude-api-key  (symmetric; we use read-modify-write PUT for all mutations)
import { log } from "../utils/logger"

const PROXY_BASE_URL = process.env.PROXY_BASE_URL || "http://127.0.0.1:8317"
const MGMT_BASE = `${PROXY_BASE_URL.replace(/\/+$/, "")}/v0/management`
const TIMEOUT_MS = 15_000

function secret(): string {
  return process.env.MANAGEMENT_SECRET || ""
}

// ── Wire types (CLIProxyAPI config shape — kebab-case keys) ──────────────────

export interface ProviderModel {
  name: string
  alias?: string
}

export interface OpenAIKeyEntry {
  "api-key": string
  "proxy-url"?: string
}

export interface OpenAICompatProvider {
  name: string
  "base-url": string
  "api-key-entries": OpenAIKeyEntry[]
  models: ProviderModel[]
  headers?: Record<string, string>
  prefix?: string
  disabled?: boolean
}

export interface ClaudeKeyProvider {
  "api-key": string
  "base-url": string
  "proxy-url"?: string
  models: ProviderModel[]
}

export class ManagementError extends Error {
  constructor(message: string, readonly status: number) {
    super(message)
    this.name = "ManagementError"
  }
}

export class ManagementUnavailableError extends Error {
  constructor(message: string) {
    super(message)
    this.name = "ManagementUnavailableError"
  }
}

// ── Low-level fetch ──────────────────────────────────────────────────────────

async function mgmtFetch(path: string, init: RequestInit = {}): Promise<Response> {
  const sk = secret()
  if (!sk) {
    throw new ManagementUnavailableError(
      "MANAGEMENT_SECRET is not set — cannot reach the CLIProxyAPI management API. " +
        "Add remote-management.secret-key to the proxy config and set MANAGEMENT_SECRET to match.",
    )
  }
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS)
  try {
    return await fetch(`${MGMT_BASE}${path}`, {
      ...init,
      headers: {
        Authorization: `Bearer ${sk}`,
        "Content-Type": "application/json",
        ...(init.headers ?? {}),
      },
      signal: controller.signal,
    })
  } catch (e) {
    throw new ManagementUnavailableError(
      `CLIProxyAPI management API unreachable at ${MGMT_BASE}: ${e instanceof Error ? e.message : String(e)}`,
    )
  } finally {
    clearTimeout(timer)
  }
}

// NOTE: never include the raw proxy response body in an error that reaches the
// client — the PUT body (and possibly echoed errors) contains plaintext API keys.
// Log the full body server-side; throw a generic, key-free message.
async function mgmtGet<T>(path: string, key: string): Promise<T[]> {
  const res = await mgmtFetch(path, { method: "GET" })
  if (res.status === 401) throw new ManagementError("Management key rejected by proxy (401)", 401)
  if (res.status === 404) throw new ManagementUnavailableError(`Endpoint ${path} not found (404) — management may be disabled or unsupported by this proxy version`)
  if (!res.ok) {
    log.error("proxyAdmin", `GET ${path} → HTTP ${res.status}: ${await res.text().catch(() => "")}`)
    throw new ManagementError(`Proxy management GET ${path} failed (HTTP ${res.status})`, res.status)
  }
  let data: Record<string, unknown>
  try {
    data = (await res.json()) as Record<string, unknown>
  } catch (e) {
    log.error("proxyAdmin", `GET ${path} returned non-JSON`, e)
    throw new ManagementError(`Proxy management GET ${path} returned an unparseable response`, 502)
  }
  const list = data[key]
  return Array.isArray(list) ? (list as T[]) : []
}

async function mgmtPut(path: string, body: unknown): Promise<void> {
  const res = await mgmtFetch(path, { method: "PUT", body: JSON.stringify(body) })
  if (res.status === 401) throw new ManagementError("Management key rejected by proxy (401)", 401)
  if (!res.ok) {
    log.error("proxyAdmin", `PUT ${path} → HTTP ${res.status}: ${await res.text().catch(() => "")}`)
    throw new ManagementError(`Proxy management PUT ${path} failed (HTTP ${res.status})`, res.status)
  }
}

// ── OpenAI-compatibility providers ───────────────────────────────────────────

export function listOpenAICompat(): Promise<OpenAICompatProvider[]> {
  return mgmtGet<OpenAICompatProvider>("/openai-compatibility", "openai-compatibility")
}

export async function putOpenAICompat(list: OpenAICompatProvider[]): Promise<void> {
  await mgmtPut("/openai-compatibility", list)
  log.llm("proxyAdmin: openai-compatibility updated", `${list.length} provider(s)`)
}

// ── Claude (Anthropic-compatible) API-key providers ──────────────────────────

export function listClaudeApiKey(): Promise<ClaudeKeyProvider[]> {
  return mgmtGet<ClaudeKeyProvider>("/claude-api-key", "claude-api-key")
}

export async function putClaudeApiKey(list: ClaudeKeyProvider[]): Promise<void> {
  await mgmtPut("/claude-api-key", list)
  log.llm("proxyAdmin: claude-api-key updated", `${list.length} provider(s)`)
}

// ── Availability probe ───────────────────────────────────────────────────────

export interface ManagementStatus {
  available: boolean
  reason?: string
}

export async function getManagementStatus(): Promise<ManagementStatus> {
  if (!secret()) return { available: false, reason: "MANAGEMENT_SECRET not configured" }
  try {
    await listOpenAICompat()
    return { available: true }
  } catch (e) {
    if (e instanceof ManagementError && e.status === 401) return { available: false, reason: "Management key rejected by proxy (401)" }
    return { available: false, reason: e instanceof Error ? e.message : String(e) }
  }
}

File: server-py/src/dana/llm/proxy_admin.py
"""Typed async client for the CLIProxyAPI Management API (⇄ TS llm/proxyAdmin.ts).

  Base:  $PROXY_BASE_URL/v0/management
  Auth:  Authorization: Bearer <MANAGEMENT_SECRET>
  GET/PUT/DELETE /openai-compatibility   (wrapper key "openai-compatibility")
  GET/PUT        /claude-api-key          (read-modify-write PUT for all mutations)

PUT bodies contain plaintext API keys, so callers must never echo them back to the browser.
"""
import httpx

from ..config import settings


class ManagementUnavailable(Exception):
    """Proxy/management API unreachable or disabled."""


class ManagementError(Exception):
    def __init__(self, message: str, status: int = 500):
        super().__init__(message)
        self.status = status


def _base() -> str:
    return settings.proxy_base_url.rstrip("/") + "/v0/management"


def _secret() -> str:
    sk = settings.management_secret
    if not sk:
        raise ManagementUnavailable(
            "MANAGEMENT_SECRET is not set — cannot reach the CLIProxyAPI management API. "
            "Add remote-management.secret-key to the proxy config and set MANAGEMENT_SECRET to match."
        )
    return sk


async def _fetch(method: str, path: str, body=None) -> httpx.Response:
    headers = {"Authorization": f"Bearer {_secret()}", "Content-Type": "application/json"}
    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            return await client.request(method, f"{_base()}{path}", headers=headers, json=body)
    except httpx.HTTPError as e:
        raise ManagementUnavailable(f"CLIProxyAPI management API unreachable at {_base()}: {e}") from e


async def _get_list(path: str, key: str) -> list[dict]:
    res = await _fetch("GET", path)
    if res.status_code == 401:
        raise ManagementError("Proxy management rejected the key (401)", 401)
    if res.status_code == 404:
        raise ManagementUnavailable(f"Endpoint {path} not found (404) — management disabled or unsupported")
    if res.status_code >= 400:
        raise ManagementError(f"GET {path} failed (HTTP {res.status_code})", res.status_code)
    data = res.json()
    return data.get(key) or [] if isinstance(data, dict) else (data or [])


async def _put_list(path: str, lst: list[dict]) -> None:
    res = await _fetch("PUT", path, lst)
    if res.status_code >= 400:
        raise ManagementError(f"PUT {path} failed (HTTP {res.status_code})", res.status_code)


# ── openai-compatibility ────────────────────────────────────────────────────────
async def list_openai_compat() -> list[dict]:
    return await _get_list("/openai-compatibility", "openai-compatibility")


async def put_openai_compat(lst: list[dict]) -> None:
    await _put_list("/openai-compatibility", lst)


# ── claude-api-key ──────────────────────────────────────────────────────────────
async def list_claude_key() -> list[dict]:
    return await _get_list("/claude-api-key", "claude-api-key")


async def put_claude_key(lst: list[dict]) -> None:
    await _put_list("/claude-api-key", lst)


async def management_status() -> dict:
    """Best-effort reachability probe."""
    try:
        await list_openai_compat()
        return {"reachable": True}
    except ManagementUnavailable as e:
        return {"reachable": False, "reason": str(e)}
    except ManagementError as e:
        return {"reachable": False, "reason": str(e), "status": e.status}


# ── OAuth (interactive provider login) ⇄ TS proxyAdmin getProviderAuthUrl / getAuthStatus ──
AUTH_URL_ENDPOINT = {
    "claude": "anthropic-auth-url",
    "openai": "codex-auth-url",
    "gemini": "gemini-cli-auth-url",
}


async def _get_json(path: str) -> dict:
    res = await _fetch("GET", path)
    if res.status_code == 401:
        raise ManagementError("Proxy management rejected the key (401)", 401)
    if res.status_code == 404:
        raise ManagementUnavailable(f"Endpoint {path} not found (404) — management disabled or unsupported")
    if res.status_code >= 400:
        raise ManagementError(f"GET {path} failed (HTTP {res.status_code})", res.status_code)
    try:
        data = res.json()
    except Exception as e:  # noqa: BLE001 — non-JSON body, ⇄ TS mgmtJson 502
        raise ManagementError(f"GET {path} returned non-JSON", 502) from e
    return data if isinstance(data, dict) else {}


async def get_provider_auth_url(endpoint: str) -> dict:
    """⇄ getProviderAuthUrl: GET /{anthropic|codex|gemini-cli}-auth-url → {url, state}."""
    return await _get_json(f"/{endpoint}")


async def get_auth_status(state: str) -> dict:
    """⇄ getAuthStatus: GET /get-auth-status?state=<s> → {status}."""
    from urllib.parse import quote
    return await _get_json(f"/get-auth-status?state={quote(state, safe='')}")
